Clustering active nodes

vkhatale opened the pull request:
[STRMALERT-1813]. Integrated missing alerts with cluster manager.

Commit:
[STRMALERT-1813]. Integrated missing alert execution with cluster manager.

Commit:
Merge branch 'clustering' into clustering_active_nodes

Commit:
[STRMALERT-1813]. Adding Missing handler

Commit:
Fixed missing alerts integration

Commit:
[STRMALERT-1813]. Fixed bugs related to missing alerts

Commit:
[STRMALERT-1813]. Refactored code to get all alerts only once when executing missing alerts.

sanchitraizada commented on pkg/services/clustering/cluster_manager.go:
```diff
@@ -259,7 +294,27 @@ func (cm *ClusterManager) handleAlertRulesDispatcherTask(task *DispatcherTask) {
 }
 
+func (cm *ClusterManager) processMissingAlerts() {
+	missingAlerts := cm.hasMissingAlerts()
```
I am not sure how expensive is the fetching the list of missing alerts (missing_alert_check can linearly depend on the number of alerts).  

If all the nodes have to compute this before one of them picks up, then I would reverse the order
i.e. first elect a leader node (candidate) for missing alert
then let only this node compute missing alerts.

Commit:
[STRMALERT-1813]. Updated code to process missing alerts after getting dispatcher status and updating the alert status in cluster manager. Also Removed unique constraint on node_id, heartbeat columns as the insertion would fail when inserting heartbeat while processing missing alerts.

Commit:
CheckIn() updates the partitionId incrementally but for missing alerts we want partitionId=0 always, so calling CheckInNodeProcessingMissingAlerts(). Also added test case to getNodeProcessingMissingAlerts

Commit:
Added unit test for processing missing alerts in cluster manager.

Commit:
Updated unique index.
1. 1. node_id,hearbeat,run_Type (to take care of nodes going down and come backup in same interval)
2.heartbeat, part_id, alert_run_type(to take care of missing alerts and different nodes having same partition id)

sanchitraizada reviewed (approved):
Generally looks ok  but few points

1. Maybe, we should have an ability to disable or enable missing alert functionality.
This is because let's say, there is a huge pileup of missing alerts (because of a "known" maintenance or planned outage) and we don't want to put resources to process missing alerts in past (and rather concentrate on real time).
2. Another potential issue, we can see is that how do we isolate the consecutive run for missing alerts - i.e. Let's says there are 4 node cluster -
(a) run from time T1, Node N1 is picked for missing alert (which still continues for many minutes) and 
(b). next minute T2, there is another run for missing alerts which can be scheduled. How is this run going to work?

Commit:
Merge branch 'clustering' into clustering_active_nodes

# Conflicts:
#	pkg/services/clustering/cluster_manager.go

vkhatale commented:
@sanchitraizada : Agree with point 1.
Point 2(b) wont be a problem because the checkin ticker runs every minute so N1 will register itself with “N1, T2, missing, processing”  (node_id, heartbeat, run_type, alert_status).  As there is any entry for N1 already processing missing no other node would pick up.
